File: source/bindings/csharp/conversation_transcription_result.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
using System;
using System.Globalization;
using Microsoft.CognitiveServices.Speech.Internal;

namespace Microsoft.CognitiveServices.Speech.Transcription
{
    /// <summary>
    /// Class that defines the result of conversation transcriber.
    /// Added in version 1.5.0
    /// </summary>
    public sealed class ConversationTranscriptionResult : SpeechRecognitionResult
    {
        internal ConversationTranscriptionResult(IntPtr resultPtr)
            : base(resultPtr)
        {
            UserId = SpxFactory.GetDataFromHandleUsingDelegate(Internal.ConversationTranscriptionResult.conversation_transcription_result_get_user_id, resultHandle, maxCharCount);
            UtteranceId = SpxFactory.GetDataFromHandleUsingDelegate(Internal.ConversationTranscriptionResult.conversation_transcription_result_get_utterance_id, resultHandle, maxCharCount);
        }

        /// <summary>
        /// A string that represents the user id.
        /// </summary>
        public string UserId { get; }

        /// <summary>
        /// A string that represents the utterance. This id is consistence for intermediates and final speech recognition result from one speaker.
        /// </summary>
        public string UtteranceId { get; }

        /// <summary>
        /// Returns a string that represents the conversation transcription result.
        /// </summary>
        /// <returns>A string that represents the conversation transcription result.</returns>
        public override string ToString()
        {
            return string.Format(CultureInfo.InvariantCulture, "ResultId:{0} Reason:{1} UserId:<{2}> UtteranceId:<{3}> Recognized text:<{4}> ",
                ResultId, Reason, UserId, UtteranceId, Text);
        }
    }
}

File: source/bindings/java/com/microsoft/cognitiveservices/speech/transcription/ConversationTranscriptionResult.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
package com.microsoft.cognitiveservices.speech.transcription;

import com.microsoft.cognitiveservices.speech.util.Contracts;
import com.microsoft.cognitiveservices.speech.util.SafeHandle;
import com.microsoft.cognitiveservices.speech.SpeechRecognitionResult;
import com.microsoft.cognitiveservices.speech.util.StringRef;

/**
 * Class that defines conversation transcription result.
 * Added in version 1.5.0
 */
public class ConversationTranscriptionResult extends SpeechRecognitionResult {

    /*! \cond PROTECTED */

    protected ConversationTranscriptionResult(long result) {
        super(result);

        if (result != 0) {
            StringRef userIdStr = new StringRef("");
            Contracts.throwIfFail(getUserId(resultHandle, userIdStr));
            this.userId = userIdStr.getValue();

            StringRef utteranceIdStr = new StringRef("");
            Contracts.throwIfFail(getUtteranceId(resultHandle, utteranceIdStr));
            this.utteranceId = utteranceIdStr.getValue();
        }
    }

    /*! \endcond */

    /**
     * Explicitly frees any external resource attached to the object
     */
    @Override
    public void close() {
        super.close();
    }

    /**
     * A String that represents the user id in the conversation.
     * @return the user ID string.
     */
    public String getUserId() {
        return this.userId;
    }

    /**
     * A string that represents the utterance. This id is consistence for intermediates and final speech recognition result from one speaker.
     * @return the utterance ID string.
     */
    public String getUtteranceId() {
        return this.utteranceId;
    }

    /**
     * Returns a String that represents the conversation transcription result.
     * @return A String that represents the conversation transcription result.
     */
    @Override
    public String toString() {
        return "ResultId:" + this.getResultId()+
                " Status:" + this.getReason() +
                " UserId:" + this.userId +
                " UtteranceId:" + this.utteranceId +
                " Recognized text:<" + this.getText() + ">.";
    }

    private final native long getUserId(SafeHandle resultHandle, StringRef userId);
    private final native long getUtteranceId(SafeHandle resultHandle, StringRef utteranceId);

    private String userId;
    private String utteranceId;
}
